src/OpenSpaceToolkit/Astrodynamics/Trajectory/Segment.cpp

```diff
@@ -584,38 +622,269 @@ const NumericalSolver& Segment::accessNumericalSolver() const
+                                                                const Interval& aManeuverInterval, const String& aSuffix
+                                                            ) -> Shared<Thruster>
+    {
+        const Shared<HeterogeneousGuidanceLaw> heterogeneousGuidanceLaw = std::make_shared<HeterogeneousGuidanceLaw>();
+
+        heterogeneousGuidanceLaw->addGuidanceLaw(originalGuidanceLaw, aManeuverInterval);
+
+        return std::make_shared<Thruster>(
+            segmentThrusterDynamics->getSatelliteSystem(),
+            heterogeneousGuidanceLaw,
+            segmentThrusterDynamics->getName() + " " + aSuffix
+        );
+    };
+
+    // Helper lambda to build a coast segment
+    const auto buildCoastSegment = [&](const String& aSuffix) -> Segment
+    {
+        return Segment::Coast(name_ + " " + aSuffix, eventCondition_, freeDynamicsArray_, numericalSolver_);
+    };
+
+    // Helper lambda to build a maneuver segment
+    const auto buildManeuverSegment = [&](const Shared<Thruster>& aThrusterDynamics, const String& aSuffix) -> Segment
+    {
+        return Segment(
+            name_ + " " + aSuffix,
+            Segment::Type::Maneuver,
+            eventCondition_,
+            freeDynamicsArray_,
+            aThrusterDynamics,
+            numericalSolver_,
+            constantManeuverDirectionLocalOrbitalFrameFactory_,
+            constantManeuverDirectionMaximumAllowedAngularOffset_
+        );
+    };
+
+    // Account for maneuver separation w.r.t. the last maneuver in the sequence
+    if (lastManeuverInterval.isDefined() && maneuverConstraints_.minimumSeparation.isDefined() &&
+        (aState.accessInstant() - lastManeuverInterval.getEnd()) < maneuverConstraints_.minimumSeparation)
+    {
+        nextInstantToCoastTo =
+            updateNextCoastInstant(lastManeuverInterval.getEnd() + maneuverConstraints_.minimumSeparation);
+        nextSubSegmentToSolve = buildCoastSegment("(Coast - Minimum Maneuver Separation Constraint)");
+    }
+
+    // TBI: is this necessary?
+    const Duration subSegmentManeuverMargin = Duration::Seconds(10.0);
+
+    while (segmentStates.accessLast().accessInstant() < maximumInstant)
+    {
+        // Coast
+        if (nextSubSegmentToSolve.getType() == Segment::Type::Coast)
+        {
+            const Segment::Solution coastSegmentSolution = nextSubSegmentToSolve.solve(
+                segmentStates.accessLast(),
+                std::min(nextInstantToCoastTo, maximumInstant) - segmentStates.accessLast().accessInstant()
+            );
+            subSegmentConditionIsSatisfied = coastSegmentSolution.conditionIsSatisfied;
+
+            segmentStates.add(Array<State>(coastSegmentSolution.states.begin() + 1, coastSegmentSolution.states.end()));
+
+            if (subSegmentConditionIsSatisfied)
+            {
+                break;
+            }
+
+            nextSubSegmentToSolve = *this;
```

Comment:
Can you put a comment here explaining why we need to resolve this subsgment if the condition isn't satisfied? and why that doesn't lead to an infinite loop? Is it because segmentStates have been added compared to the last time this ran - so you have still made progress towards the end of the segment?